File: csharp/1654.minimum-jumps-to-reach-home.cs
/*
 * @lc app=leetcode id=1654 lang=csharp
 *
 * [1654] Minimum Jumps to Reach Home
 */

using System;
using System.Collections.Generic;
using System.Linq;

// @lc code=start
public class Solution {

    private static int MinimumJumps(HashSet<int> forbidden, int a, int b, int x, int lim, int curr, bool forward, HashSet<int> seen, IDictionary<(int Index, bool Forward), int> memo) {
        if (curr == x) {
            return 0;
        }
        var key = (Index: curr, Forward: forward);
        if (memo.TryGetValue(key, out var ret)) {
            return ret;
        }
        ret = int.MaxValue;
        int f = curr + a, k = curr - b;
        seen.Add(curr);
        if (f <= lim && !forbidden.Contains(f) && !seen.Contains(f)) {
            int fr = MinimumJumps(forbidden, a, b, x, lim, f, true, seen, memo);
            if (fr < int.MaxValue) {
                ret = 1 + fr;
            }
        }
        if (forward && k > 0 && !forbidden.Contains(k) && !seen.Contains(k)) {
            int kr = MinimumJumps(forbidden, a, b, x, lim, k, false, seen, memo);
            if (kr < int.MaxValue) {
                ret = Math.Min(1 + kr, ret);
            }
        }
        seen.Remove(curr);
        return memo[key] = ret;
    }

    public int MinimumJumps(int[] forbidden, int a, int b, int x) {
        int ret = MinimumJumps(new HashSet<int>(forbidden), a, b, x, forbidden.Append(x).Max() + a * 2 + b * 2, 0, true, new HashSet<int>(), new Dictionary<(int Index, bool Forward), int>());
        return ret == int.MaxValue ? -1 : ret;
    }
}
// @lc code=end

File: java/1654.minimum-jumps-to-reach-home.java
/*
 * @lc app=leetcode id=1654 lang=java
 *
 * [1654] Minimum Jumps to Reach Home
 */

import java.util.ArrayDeque;
import java.util.Queue;

// @lc code=start

class Solution {
    public int minimumJumps(int[] forbidden, int a, int b, int x) {
        int gcd = gcd(a, b);
        if (x % gcd != 0) {
            return -1;
        }
        int ret = 0;
        boolean[] forbiddenB = new boolean[2001];
        for (int f : forbidden) {
            forbiddenB[f] = true;
        }
        boolean[][] visited = new boolean[10001][2];
        Queue<Pair> queue = new ArrayDeque<>();
        queue.offer(new Pair(0, true));
        while (!queue.isEmpty()) {
            int k = queue.size();
            while (k-- > 0) {
                Pair p = queue.poll();
                if (p.curr() == x) {
                    return ret;
                }
                int f = p.curr() + a;
                if (f <= 10000 && (f > 2000 || !forbiddenB[f]) && !visited[f][1]) {
                    queue.offer(new Pair(f, true));
                    visited[f][1] = true;
                }
                if (p.forward()) {
                    int back = p.curr() - b;
                    if (back > 0 && (back > 2000 || !forbiddenB[back]) && !visited[back][0]) {
                        queue.offer(new Pair(back, false));
                        visited[back][0] = true;
                    }
                }
            }
            ++ret;
        }
        return -1;
    }

    private static int gcd(int a, int b) {
        return b == 0 ? a : gcd(b, a % b);
    }

    private record Pair(int curr, boolean forward) {}
}
// @lc code=end

